Fix Nuage provider update

sasoc opened the pull request:
There were two problems when editing Nuage network provider. If
network that you edited had no event, there would still be AMQP
selected under the Events tab. That made Save button disabled
since hostname must not be empty and everything has to be verified.
Now None is selected in case of no event.
The second problem was that when a network was updated error occurred
while redirecting, since you would only get one argument and
`ems_path` required two. Now `ems_path` works same as
in the cloud provider controller.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1534457

sasoc commented:
cc @miha-plesko

miha-plesko commented on app/controllers/mixins/ems_common_angular.rb:
```diff
@@ -713,8 +713,13 @@ def build_credentials(ems, mode)
+        return 'none' if amqp_auth.nil?
+        return 'amqp'
+      end
+      'ceilometer'
```
I don't like that 'ceilometer' is fallback when nothing else is true, can you please reorganize function to following shape:

```
return 'amqp' if ...
return 'ceilometer' if ...
return 'none'
```

miha-plesko replied:
Also, please add unit test for this function, e.g. in a new file `spec/controllers/mixins/ems_common_angular_spec.rb`.

miha-plesko commented on app/controllers/mixins/ems_common_angular.rb:
```diff
@@ -713,8 +713,10 @@ def build_credentials(ems, mode)
-      return "amqp" if @ems.connection_configurations.ceilometer.try(:endpoint).nil? && @ems.connection_configurations.amqp.try(:endpoint)
-      "ceilometer"
+      amqp_auth = @ems.authentications.detect { |a| a.authtype == 'amqp' }
+      return 'amqp' if @ems.connection_configurations.ceilometer.try(:endpoint).nil? && @ems.connection_configurations.amqp.try(:endpoint) && !amqp_auth.nil?
```
```ruby
return 'amqp' if @ems.connection_configurations.amqp&.endpoint&.hostname&.present?
```

miha-plesko replied:
I don't think we need to check both auth and endpoint, it should suffice to have a valid hostname where to connect to, please see example above.

miha-plesko commented on app/controllers/mixins/ems_common_angular.rb:
```diff
@@ -713,8 +713,10 @@ def build_credentials(ems, mode)
-      "ceilometer"
+      amqp_auth = @ems.authentications.detect { |a| a.authtype == 'amqp' }
+      return 'amqp' if @ems.connection_configurations.ceilometer.try(:endpoint).nil? && @ems.connection_configurations.amqp.try(:endpoint) && !amqp_auth.nil?
+      return 'ceilometer' if @ems.connection_configurations.ceilometer.try(:endpoint) && @ems.connection_configurations.amqp.try(:endpoint).nil?
```
```ruby
return 'ceilometer' if @ems.connection_configurations.ceilometer&.endpoint&.hostname&.present?
```

miha-plesko commented on spec/controllers/mixins/ems_common_angular_spec.rb:
```diff
@@ -0,0 +1,27 @@
+    let(:ems_nuage) { FactoryGirl.create(:ems_nuage_network_with_authentication) }
+
+    it 'when amqp' do
+      ems_nuage.endpoints << Endpoint.create(:role => 'amqp')
```
`Endpoint.create(:role => 'amqp', hostname => 'hostname')`

miha-plesko commented on spec/controllers/mixins/ems_common_angular_spec.rb:
```diff
@@ -0,0 +1,27 @@
+      expect(network_controller.send(:retrieve_event_stream_selection)).to eq('none')
+    end
+  end
+end
```
Can you please add also following tests:

* AMQP when ceilometer has empty hostname
* CEILOMETER when amqp has empty hostname
* NONE when amqp and ceilometer have nil endpoints
* NONE when amqp and cailometer have empty hostnames

miha-plesko commented on spec/controllers/mixins/ems_common_angular_spec.rb:
```diff
@@ -0,0 +1,57 @@
+      expect(network_controller.send(:retrieve_event_stream_selection)).to eq('amqp')
+    end
+
+    it 'none when amqp and ceilometer have nil endpoints' do
```
This is the very same as we already have in L51, please remove it.

miha-plesko commented:
@AparnaKarve I kindly ask for a review/merge. The problem that we're addressing here is blocking integration tests for Nuage provider, because simple CRUD operations fail due to bug on the UI. Bug was there for a long time now, but no one noticed...

@miq-bot assign @AparnaKarve 
@miq-bot add_label enhancement

/cc @juliancheal

AparnaKarve commented:
@miha-plesko Need a little clarification here.
Is this PR applicable to Nuage Network Providers only? 
Or is it applicable to Nuage and Amazon? ( I believe in the BZ you have mentioned Amazon)

Incidentally, I can open the edit page for Amazon Network Provider (even without this PR)
But the form does not populate all the fields. 

<img width="1431" alt="screen shot 2018-02-01 at 6 44 52 pm" src="https://user-images.githubusercontent.com/1538216/35714210-32ed97c2-0780-11e8-82e7-3343bd5ca72c.png">

miha-plesko commented:
@AparnaKarve I think the "Edit NetworkProvider" option should be disabled for any other provider but Nuage. Probably we've accidentaly enabled it for all providers when we were integrating Nuage. 

We've also tested if editing works for Amazon NetworkManager but it didn't (it's broken on many places, as you've correctly noticed) therefore we were brave enough to assume `def ems_path(*args)` in ems_network_controller.rb is obsolete because no one has used it for a while now - hence this PR.

May I suggest following procedure:

- this PR: make Nuage NetworkManager editing work
- followup PR: disable editing for all other NetworkManagers but Nuage

Wdyt?

P.S.: Not sure how familiar are you with Nuage provider, so let me quickly put some context for you. Nuage Provider is "one of a kind" among providers because it no CloudManager, only NetworkManager. Other providers (like Amazon) only allow you to add a new CloudProvider and then Network-, Storage-, ... managers are added automatically (you are not even able to edit them). When you remove Amazon CloudProvider, also all its Managers are removed. Well, this is not the case with Nuage, where there is no CloudProvider, only a standalone NetworkManager -> and that's the source of our problem, because UI does not assume Manager editing. Luckily, the fixes seem to be quite simple 😄 

Please let me know if this makes any sense.

AparnaKarve commented on app/controllers/mixins/ems_common_angular.rb:
```diff
@@ -713,8 +713,9 @@ def build_credentials(ems, mode)
-      "ceilometer"
+      return 'amqp' if @ems.connection_configurations.amqp&.endpoint&.hostname&.present?
+      return 'ceilometer' if @ems.connection_configurations.ceilometer&.endpoint&.hostname&.present?
+      'none'
```
Noticed that a `hostname&.present?` condition has been added which did not exist before.
Should not matter for OpenStack Cloud/Infra Provider I think (where this method is used), but just wanted to know if it's absolutely needed for Nuage?

AparnaKarve replied:
Could you post a screenshot of the Nuage Edit Network Provider page?
Does it have AMQP/Ceilometer Radio buttons too?

AparnaKarve replied:
I think the above change may affect how we display the Radio buttons for OpenStack.

OpenStack selects the Ceilometer radiobutton by default if the AMQP endpoint does not exist.
But with the above change, there is an explicit check for the Ceilometer endpoint.
If both AMQP and Ceilometer endpoints do not exist, none of the radiobuttons would be selected -- which is not what we want.

Let's keep the OpenStack UI unaffected by this change.

AparnaKarve commented:
@miha-plesko Thanks for the detailed explanation above - it did make sense :)

I have a few inline comments.

sasoc commented:
@AparnaKarve `hostname&.present?` was added because there is a bug when you create new Nuage network. When created without endpoints there is still endpoint added to network, which has empty attributes. And before update `retrieve_event_stream_selection` would return `'amqp'` and amqp radio button would be selected in edit page. And since all attributes are empty and
form fields are required, you would have to go to events tab and select none so you would be allowed to save changes(as seen in screenshot). Which creates some unnecessary work. 

I've changed `retrieve_event_stream_selection` to return `ceilometer` by default for OpenStack. I've also added test for this usecase.

Screenshot of Nuage network edit page, where you can see that there is oly AMQP radio button:
<img width="925" alt="screen shot 2018-02-06 at 10 39 31" src="https://user-images.githubusercontent.com/9213557/35855251-f225f656-0b32-11e8-87a0-59c9891676f2.png">

AparnaKarve commented:
@sasoc @miha-plesko This looks good now.

@dclarizio Can we merge this?

AparnaKarve commented:
@miha-plesko Regarding the follow-up PR as mentioned in https://github.com/ManageIQ/manageiq-ui-classic/pull/3337#issuecomment-362512635 -- is that something you would be able to address?
Let us know. Thanks.

miha-plesko commented:
Thanks for merging. @AparnaKarve  yes, let me prepare followup PR for you. I think it should be a matter of a single IF statement, I'll take a look on Monday.

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit 7bc1af0116223f5ff9eb5aa239ac4b0ba77b9621
Author: Dan Clarizio <dclarizi@redhat.com>
Date:   Fri Feb 9 14:52:50 2018 -0800

    Merge pull request #3337 from sasoc/fix-nuage-provider-update
    
    Fix Nuage provider update
    (cherry picked from commit 02e1a7098b58cff9b7cc309b18e5fda7fe56e17e)
    
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1552895
```
